File: 2100-find-good-days-to-rob-the-bank/2100-find-good-days-to-rob-the-bank.py
class Solution:
    def goodDaysToRobBank(self, security: List[int], time: int) -> List[int]:
        n = len(security)
        inc_cnt = [0 for i in range(n)]
        dec_cnt = [0 for i in range(n)]
        ans = []
        
        for i in range(1,n):
            if security[i] <= security[i - 1]:
                dec_cnt[i] = 1 + dec_cnt[i - 1]
        
        for j in range(n - 2, -1, -1):
            if security[j] <= security[j + 1]:    
                inc_cnt[j] = inc_cnt[j + 1] + 1 
      
        for i in range(n):
            if inc_cnt[i] >= time and dec_cnt[i] >= time:
                ans.append(i)
                
        return ans
            
            
        

File: 2100-find-good-days-to-rob-the-bank/2100-find-good-days-to-rob-the-bank.java
class Solution {
    private static int dp(int i, int dr, Map<String, Integer> memo, int n, int [] security){
        String key = i + " , " + dr;
        if (memo.containsKey(key)) return memo.get(key);
        if (0 <= i + dr && i + dr < n &&  security[i  + dr] >= security[i]){
            memo.put(key ,1 + dp(i + dr, dr, memo, n, security));
            return memo.get(key);
        }
        return 0;
    }
    public List<Integer> goodDaysToRobBank(int[] security, int time) {
        List<Integer> ans = new ArrayList<Integer>();
        int n = security.length;
        Map<String, Integer> memo = new HashMap<>();
        for(int i = 0; i < n; i++){
            int first = dp(i, 1, memo, n, security);
            int second = dp(i, -1, memo, n, security);
            if (first >= time && second >= time){
                ans.add(i);
            }
        }
        return ans;
    }
}